history MD5 hash calculation

phweyland opened the pull request:
Start for #4313 

History hash calculation
Display of hash on image information (just for debug)
Time displayed on console (hovering images)

```
hash time 0,000009
get history plus hash time 0,000957
hash time 0,000008
get history plus hash time 0,000557
hash time 0,000019
get history plus hash time 0,000614
hash time 0,000032
get history plus hash time 0,002047
hash time 0,000035
get history plus hash time 0,002301
hash time 0,000032
get history plus hash time 0,002467
```

I've added a routine to get the history copying some pieces of code without understanding every steps. Creating a dev context just to get history seems very heavy.
If hash calculation is made on dev->history when available this is not an issue. Otherwise I need to go deeper to determine the minimum necessary to get whatever image history.

phweyland commented:
> Otherwise I need to go deeper to determine the minimum necessary to get whatever image history

So far I understand that the history is available in db except the iop_order.
Then some additional calculation must be done to complete the iop_order for each active module. Is that correct ?

TurboGit commented:
> Creating a dev context just to get history seems very heavy.

Yes, I would certainly write a dedicated SQL query to get the enabled module only on the history with the needed information. This routine would return a GList*.

`SELECT operation, op_params, blendop_params FROM history WHERE enabled = 1 AND imgid = ?1 ORDER BY num;`

Note the `ORDER BY num` which ensure that the modules are always compared in the right order. I don't think we need the multi_priority. After all if you duplicate an instance, delete the original one you still have the same development, no?

TurboGit commented:
> So far I understand that the history is available in db except the iop_order.

Right, we need to include into the hash the version and iop_list from module_order I suppose. This with the history as proposed above should be enough to detect an altered status.

phweyland commented:
> After all if you duplicate an instance, delete the original one you still have the same development, no?

Hummm, multi multipriority = multi instance.
If so, each instance, added or removed changes the image, doesn't it ?

TurboGit commented:
> If so, each instance, added or removed changes the image, doesn't it ?

But without this value in the history you'll have one or two instance. The hash will be different anyway.

phweyland commented:
> Right, we need to include into the hash the version and iop_list from module_order I suppose. This with the history as proposed above should be enough to detect an altered status.

OK, that would be a way, instead of going through the iop_order calculation.

phweyland commented:
> > Right, we need to include into the hash the version and iop_list from module_order I suppose. This with the history as proposed above should be enough to detect an altered status.
> 
> OK, that would be a way, instead of going through the iop_order calculation.

I don't think that's a good idea ... most of the work will be done on dev phase (on dev->history).
So to make the comparison we need to calculate the checksum on the same structure.

If that's correct, whatever I do, I need to mimic exactly the dev_history list (at least with the needed parameters).

phweyland commented:
> > If so, each instance, added or removed changes the image, doesn't it ?
> 
> But without this value in the history you'll have one or two instance. The hash will be different anyway.

Right.

TurboGit commented:
> > 
> 
> I don't think that's a good idea ... most of the work will be done on dev phase (on dev->history).
> So to make the comparison we need to calculate the checksum on the same structure.

You don't need that (or not required). From lightable you don't have a dev struct and from the darkroom the calculation would be done after writing the history so all is in the DB.

TurboGit commented:
My idea is that the hash must be *content oriented* and not *topology oriented*.

phweyland commented:
> the calculation would be done after writing the history so all is in the DB

In that case that's ok.

TurboGit commented on src/common/history.c:
```diff
@@ -1206,6 +1206,57 @@ GList *dt_history_duplicate(GList *hist)
+
+  while(sqlite3_step(stmt) == SQLITE_ROW)
+  {
+    if(sqlite3_column_int(stmt, 2))
```
Why not put this in the SQL statement above? Looping around all non enabled modules are time consuming for nothing.

phweyland replied:
Right. I'll do.

phweyland commented:
Here is the direct calculation against db. More efficient than through the dev context creation but far slower than on in memory history list.

```
get history plus hash img 56054 time 0,000111
get history plus hash img 56047 time 0,000212
get history plus hash img 56047 time 0,000202
get history plus hash img 56054 time 0,000248
get history plus hash img 56061 time 0,000220
get history plus hash img 56068 time 0,000221
get history plus hash img 56075 time 0,000218
get history plus hash img 56082 time 0,000148
get history plus hash img 56083 time 0,000052
get history plus hash img 56084 time 0,000059
```

TurboGit commented:
> Here is the direct calculation against db. More efficient than through the dev context creation but far slower than on in memory history list.

For sure, but we can use only the dev context for darkroom where the speed (only one image at a time) is not that an issue.

Maybe with the enabled on SQL we can gain some CPU cycles?

phweyland commented:
> Maybe with the enabled on SQL we can gain some CPU cycles?

Already done.

Ok, I'll provide the same calculation from dev_history_list.

TurboGit commented:
> Already done.

So you did not push last version ?

phweyland commented:
> > Already done.
> 
> So you did not push last version ?

Sorry, my mistake. I hadn't seen your comment on the code.

TurboGit commented:
Anyway I should give you more time :) I'm looking at this as soon as I see a change :) So eager to see all this stuff in action :)

phweyland commented:
I've tried to simulate the same calculation from dev (creating a mininum context as in the first commit) but I fail on params_size calculation. 

Last commit with enabled in where condition. Maybe a bit better.

phweyland commented:
> I fail on params_size calculation.

By the way the time difference may be shared between db accesses and params hash I may have missed with the first commit.

phweyland commented:
I think we have 3 hashes to store somewhere:
- creation time / default profile - calculated
- last darkroom leave (or last user change ?) - calculated
- last thumbnail refresh - copied

To store them we could add those fields in main.images ? main.module_order ? A specific table ?

What do you think ?

Just a note: running some tests I've seen that images on which I'd copied the dev from another image share the same hash.

phweyland commented:
> My idea is that the hash must be **content oriented** and not **topology oriented**.

I don't understand this. Could you elaborate ?

TurboGit commented:
@phweyland : By **content oriented** I mean that only the content of the data is meaningful and for example we don't care about the multi_priority. The fact that expose is on a multi_priiority of 1 or 2 or whatever should not be taken into account, the important point is that the actual content (op_params and blendop_params) is used for the hash. Another way to see that is, an exposure module with a given set of parameters will have the very same effect whether it is on a multi_priority of 1 or 2... I hope I'm clear :) And I hope this is correct regarding the issue we are trying to solve.

TurboGit commented:
Can you tell me more about:
> last thumbnail refresh - copied

Does that means that you also want to replace the current dt_image_altered() routine ? Or do you have something else in mind?

TurboGit commented on src/common/history.c:
```diff
@@ -1206,6 +1206,54 @@ GList *dt_history_duplicate(GList *hist)
+  {
+    char *buf = (char *)sqlite3_column_text(stmt, 0);
+    if(buf) g_checksum_update(checksum, (const guchar *)buf, -1);
+    buf = (char *)sqlite3_column_text(stmt, 1);
```
fields 1 and 2 are not text. These are blob and the size if given by reading the first bytes:

```
    const int32_t params_len = sqlite3_column_bytes(stmt, 1);
    const void *params_blob = sqlite3_column_blob(stmt, 1);
```

Then you can use:

`
   if(params_blob) g_checksum_update(checksum, (const guchar *)params_blob, params_len);`

phweyland replied:
I've missed that. I'll fix it.

TurboGit replied:
Thanks ! And then redo the timing as with this code the checksum update was ending at the first \0 found in the blob.

phweyland commented:
> Does that means that you also want to replace the current dt_image_altered() routine ?

What does this routine ?
It returns true if it exists another module than the basic one... (I'm a bit unsure about the condition ` num<history_end`). Nearly all the time ? Is that correct ?

I think we need to refresh the thumbnail iff the image has changed since the last refresh. 
I don't think we need to refresh the thumbnails all the time.

I may have misunderstood everything ... But if I'm correct we should replace this routine too.

> Another way to see that is, an exposure module with a given set of parameters will have the very same effect whether it is on a multi_priority of 1 or 2...

understood and agreed.

TurboGit commented on src/common/history.c:
```diff
@@ -1206,6 +1206,58 @@ GList *dt_history_duplicate(GList *hist)
+  while(sqlite3_step(stmt) == SQLITE_ROW)
+  {
+    char *buf = (char *)sqlite3_column_text(stmt, 0);
+    int32_t params_len = sqlite3_column_bytes(stmt, 0);
```
No need here as operation is TEXT.

phweyland replied:
my bad ...

TurboGit commented on src/common/history.c:
```diff
@@ -1206,6 +1206,58 @@ GList *dt_history_duplicate(GList *hist)
+    char *buf = (char *)sqlite3_column_text(stmt, 0);
+    int32_t params_len = sqlite3_column_bytes(stmt, 0);
+    if(buf) g_checksum_update(checksum, (const guchar *)buf, params_len);
+    buf = (char *)sqlite3_column_text(stmt, 1);
```
This one should be sqlite3_column_blob() and not sqlite3_column_text()

phweyland replied:
all wrong ...

TurboGit commented on src/common/history.c:
```diff
@@ -1206,6 +1206,58 @@ GList *dt_history_duplicate(GList *hist)
+    buf = (char *)sqlite3_column_text(stmt, 1);
+    params_len = sqlite3_column_bytes(stmt, 1);
+    if(buf) g_checksum_update(checksum, (const guchar *)buf, params_len);
+    buf = (char *)sqlite3_column_text(stmt, 2);
```
Likewise _text -> _blob

TurboGit commented on src/common/history.c:
```diff
@@ -1206,6 +1206,58 @@ GList *dt_history_duplicate(GList *hist)
+    if(version == DT_IOP_ORDER_CUSTOM)
+    {
+      const char *buf = (char *)sqlite3_column_text(stmt, 1);
+      const int32_t params_len = sqlite3_column_bytes(stmt, 1);
```
Not needed, this is really TEXT !

phweyland commented:
I hope I've fixed these stupid mistakes ... :(

Added a table history_hash. This can be changed if you think we should of course.
I've kept the hash as a string but it could set as an hexa array. Not sure we would save a lot.
Updating the table when the user leaves an image in darkroom.

phweyland commented:
Thinking loud.
I see 3 hashes:
- neutral (default module at default state)
- auto-applied
- current

Based on the comparison of hashes we should be able to split the images in these 3 categories. Remain to make some choices for images which have been created before (default rule using the current altered-not altered status).

The common action is to update the current hash on any action (darktable leave image, partial copy of history, undo, applying style, ...)

The neutral hash should be set (along with current) when
- import without xmp before auto-apply
- update of legacy_params if neutral = current

The auto-apply hash should be set
- when auto-apply if any

Remain import with xmp or full copy of history
- if xmp contains hashes just copy
- if not, default rule

From what I've seen so far these hashes should be updated in sync with module_order. So I think the best place for them should be the module_order table.

The dt_write_iop_order_list() could hold the current hash update.

Maybe, if we are able to add a context parameter to this routine we could cover the set/update of the 3 values.

Thoughts ?

phweyland commented:
blendop_params & blendop_version are not initialized for default modules.
Then if one opens an image in darkroom and exits the hash changes... (the params and blendop_params versions are not managed yet).

phweyland commented:
- in database fulfill history_hash table with altered / not altered status from history table
- collection / history and dt_image_altered() connected to history_hash (same source now)
- capture better initial / default / altered status
- capture legacy params (not tested)

TurboGit commented:
@phweyland : I would not save the hash in the XMP. Why is that needed? It can always be computed after importing and this is fast.

TurboGit commented on src/common/history.c:
```diff
@@ -1205,6 +1205,175 @@ GList *dt_history_duplicate(GList *hist)
+      values = dt_util_dstrcat(values, "?2,");
+      conflict = dt_util_dstrcat(conflict, "current_hash=?2,");
+    }
+    if(fields) fields[strlen(fields) - 1] = '\0';
```
To remove the comas? Worth a comment I would say.

phweyland replied:
Yes. I'll do.

TurboGit commented on src/common/history.h:
```diff
@@ -25,6 +25,14 @@
 
+typedef enum dt_hash_history_t
+{
+  DT_HH_INITIAL = 1 << 0,
```
I think they are all exclusive, right ? So 0, 1, 2, 3 would have been ok too. Here it looks like as we would use bit-wise operators.

TurboGit replied:
And BTW, as this is in a spec maybe more meaningful name: DT_HISTORY_HASH_...

phweyland replied:
no, we can call the hash_history_write() asking for writing whatever hashes at once (usually default and current or initial and current), so the setting needs combination.
Currently I use the same enum for as setting and status. So I can set an other enum for the status (like you suggest). Is that you would like ?

TurboGit replied:
Fine with me, I missed the possible combination of multiple hash so ignore this part.

Still HH does not read directly, maybe to you but people coming to the code will certainly be more informed with HISTORY_HASH instead of HH (especially in a .h where some values will be used outside of the history file).

phweyland replied:
Understood. I'll do it.

phweyland replied:
done

TurboGit commented on src/common/history.h:
```diff
@@ -76,6 +84,15 @@ char *dt_history_get_items_as_string(int32_t imgid);
 gboolean dt_history_check_module_exists(int32_t imgid, const char *operation);
 
+/** calculate the hash for that history from db*/
+gsize dt_hash_history_compute_from_db(const int32_t imgid, guint8 **hash);
```
convention is that all routine in one module just start with dt_<module>_ so here it should all be:
`dt_history_hash...();`

For the three new routines.

phweyland replied:
There is a dt_history_xxx, dt_dev_history_xxx, it's why I've added a dy_hash_history_xxx.

For the routine called only in one module I've seen that `dt_` prefix is replaced by `_`. So if that the subject of the remark, I should rename `dt_hash_history_compute_from_db()` to `_hash_history_compute_from_db()`. is that correct ? If not, I haven't understood the remark.

phweyland replied:
done, both.

TurboGit replied:
@phweyland ; Thanks, not that dt_dev_history exists indeed, but in develop.h which makes my point :)

TurboGit commented on src/common/history.h:
```diff
@@ -25,6 +25,14 @@
+  DT_HH_DEFAULT = 1 << 1,
+  DT_HH_CURRENT = 1 << 2,
+  DT_HH_UNKNOWN = 1 << 3
+} dt_hash_history_t;
```
type should be : dt_history_hash_t (see below)

phweyland replied:
There is a dt_history_xxx, dt_dev_history_xxx, it's why I've added a dy_hash_history_xxx.

phweyland replied:
It's look for me more consistent as is, but I can change. Let me know.

phweyland replied:
done

TurboGit commented on src/views/darkroom.c:
```diff
@@ -30,6 +30,7 @@
 #include "common/styles.h"
 #include "common/tags.h"
 #include "common/undo.h"
+#include "common/history.h"
```
looks superfluous ?

phweyland replied:
correct. I remove it.

TurboGit commented on src/common/image.c:
```diff
@@ -937,35 +939,19 @@ void dt_image_remove(const int32_t imgid)
-
-  return altered;
+  dt_hash_history_t status = dt_hash_history_get_status(imgid);
+  return status & DT_HH_CURRENT;
```
with the suggested changes in enum this would just be:
`
return status == DT_HH_CURRENT;`

TurboGit replied:
So this is good given previous comment.

phweyland commented:
> @phweyland : I would not save the hash in the XMP. Why is that needed? It can always be computed after importing and this is fast.
> 
> EDIT: I was reacting about your previous message ([#4322 (comment)](https://github.com/darktable-org/darktable/pull/4322#issuecomment-589070414)) but it seems that this is not done. So all good.

It is still to be done. The current_hash is not worth in itself I agree. But the initial and default ones must be saved to know if the image is altered or not (the xmp is saved even if the image is not altered).

phweyland commented:
Just before committed I've added the status unknown in routine `dt_hash_history_get_status()` ... And now I've an avalanche of the status because `dt_image_altered()` is called far before the hash is created. Needs to understand why.

phweyland commented:
to fulfill history_hash table I've used the collection altered/not altered rule.
But I'll change for the former `dt_image_altered()` rules. I'll use also the same to treat an xmp without hash data.
That way we should be able to recover also some default (auto) cases.

BTW I think I will rename `default` (default_hash, ...) by `auto`.

phweyland commented:
I've made the change for resolved comments, but I've let my current printf traces for my current tests.

phweyland commented:
> Just before committed I've added the status unknown in routine `dt_hash_history_get_status()` ... And now I've an avalanche of this status because `dt_image_altered()` is called far before the hash is created. Needs to understand why.

dt_view_image_expose() (refreshes mipmap) is called at the end of import process before dt_dev_read_history() (creates history). Nothing I can do about.

TurboGit commented on src/common/database.c:
```diff
@@ -1415,6 +1415,33 @@ static int _upgrade_library_schema_step(dt_database_t *db, int version)
+    const gsize checksum_len = g_checksum_type_get_length(G_CHECKSUM_MD5);
+    guint8 *checksum = calloc(1, checksum_len);
+    sqlite3_stmt *ins_stmt = NULL;
+    char *query = dt_util_dstrcat(NULL, "INSERT INTO main.history_hash "
```
The LEFT JOIN may look surprising, I would add a comment:

// Insert an history_hash entry for all images with some history already set (skip non edited images)

phweyland replied:
The left join is to get also images without history.

// Insert an history_hash entry for all images with some history already set (including  non edited images)

phweyland replied:
done

phweyland commented:
I'm close from the scope I could think of.
With my migrated db this starts to work smoothly.
With a brand new one I have 2 issues:
- removing the auto-apply doesn't work on the first import(s ?).
- more concern about the history which is not always set up ... It's set up for opened image (several times see #4232), for another one sometimes and not for the others.

phweyland commented on src/common/exif.cc:
```diff
@@ -2943,6 +2945,32 @@ int dt_exif_xmp_read(dt_image_t *img, const char *filename, const int history_on
+      }
+      else
+      {
+        dt_history_hash_write_from_history(img->id, DT_HH_CURRENT); // no choice, use the history itelf
```
I've missed to apply the former dt_image_altered() rules here

phweyland replied:
done

phweyland commented:
Applied your comments .

TurboGit commented on src/develop/develop.c:
```diff
@@ -1310,6 +1311,18 @@ void dt_dev_write_history(dt_develop_t *dev)
+{
+  sqlite3_stmt *stmt;
+  DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db),
+                              "SELECT count (*) FROM  memory.history",
```
should be COUNT(*)

I would move this into history.h and use it in many places (at least in history.c, develop.c and lightrroom.c) where we are using this exact SQL statement. Would be a nice refactoring.

TurboGit replied:
EDIT: sorry this is counting memory not main.history so forget about this.

TurboGit commented on src/libs/collect.c:
```diff
@@ -1305,18 +1306,19 @@ static void list_view(dt_lib_collect_rule_t *dr)
                    " GROUP BY altered"
                    " ORDER BY altered ASC",
-                   _("altered"),  _("not altered"), where_ext);
+                    ("initial"), _("auto applied"), _("altered"), _("none"), where_ext);
```
Probably a missing translation here _("initial") ?

phweyland replied:
Right !

TurboGit commented:
I have tested this and see some strange things:
- first after migration I have many initial_hash set to NULL
- entering in the darkrroom with an image marked as non-altered mark it as altered (icon on lighttable)

This second point is maybe related to the first one as for new edits going to darkrroom does not mark the image as altered.

phweyland commented:
> I have tested this and see some strange things:
> 
>     * first after migration I have many initial_hash set to NULL

There are 3 cases:
- initial = current => initial status
- auto = current => auto applied status
- else => altered

So, after migration you can have
- initial = auto = NULL and current = hash => altered
- initial = current = hash (the same) and auto = NULL => initial
We don't have the case auto applied from migration


>     * entering in the darkrroom with an image marked as non-altered mark it as altered (icon on lighttable)
> 

That's an issue indeed. This should not happen if the image has an history (in that case it gets one and it changes to altered). But with the hash logic we should not have any image without history. 

Except that ... I've not solved the problem risen yesterday. With as fresh db I import images and they don't get the history. I understand that what I've coded (dt_imageio_export_with_flags) is wrong and must be normal import thread (dt_image_import_internal) instead. But I haven't yet found the configuration point which makes work my wrong code. I suspect your imported db has not this configuration flag and works as my fresh db.

In that case what you see is normal (no history at import time).
To make work the hash system we need an history at any time.

phweyland commented:
The fact to have several paths to import images is not safe ...

phweyland commented:
>     * entering in the darkrroom with an image marked as non-altered mark it as altered (icon on lighttable)

Understood I think. Not related to problem exposed above. 
At import time I set a fake checksum. This not an issue for altered images. But this is an issue for not altered ones. For sure, passing through darkroom, the image gets a different checksum. 
Without talking about the potential lack of history. 

To make this part work we could get 3 status from existing images:
- initial (without history)
- auto apply  (with history but not altered - former rule)
- altered the remaining images

second point, I need
- to get an history for initial images
- calculate the correct checksum for all images.

phweyland commented:
> But I haven't yet found the configuration point which makes work my wrong code.

Understood ! With my live db, I've the setting "don't use embedded preview jpeg but half-size raw".
This forces mipmap to reload the image. This what makes work the code as is.
Now I've just to complete dt_image_import_internal() and make import work too.

phweyland commented:
I've changed a bit my mind (I've understood better the current dt behavior :) )

With the current dt the confusion comes from the fact that fresh raw images import can lead to 2 different status depending on "don't use embedded preview jpeg but half-size raw" setting:
- if not set, no history (to speed up the import process I guess). the image gets its history when enters in darkroom.
- if set, thumbnails update provides them immediately a basic history.

That's where I've been confused. And of course `not altered` is not always equivalent to `no history` !

So, here I've kept the option not to create history at import time ( "don't use embedded preview jpeg but half-size raw" not set).

But we consider images without history or with history and basic_hash = current _hash as basic.

BTW I've changed the wording `initial` by `basic`. We have 3 states:
- basic
- auto-applied
- altered

I've fixed the migration issue, providing a real hash where needed. BTW, is there any mean to use the history hash routines here instead of duplicating code ?

There may be a improvement to be done on migration (and import of image with xmp as well).
We could improve the logic of the former dt_image_altered(), splitting images in 3 categories (basic, auto applied & altered) instead of 2 (basic & altered).
Thoughts ?

parafin commented:
Why would one need to distinguish basic from auto applied category?

phweyland commented:
> Why would one need to distinguish basic from auto applied category?

Is your question related to db migration process or is it general ?

parafin commented:
I'm talking about exposed in the GUI values.

phweyland commented:
> I'm talking about exposed in the GUI values.

This have been exposed [here](https://github.com/darktable-org/darktable/issues/4313#issuecomment-587647140) in #4313 .

I would add that I share this standpoint. basic images are surely in a different state than auto-applied ones and may deserve a different attention.

parafin commented:
As you noticed, just changing "don't use embedded preview jpeg but half-size raw" option will change freshly imported images from "basic" to "auto-applied". I don't think it gives any useful information to a user.

phweyland commented:
> As you noticed, just changing "don't use embedded preview jpeg but half-size raw" option will change freshly imported images from "basic" to "auto-applied". I don't think it gives any useful information to a user.

That's the behavior of the current master. It seems that's still the case with that PR and I'll try to fix it. That was clearly my intention to have them in basic ones.

Thanks for pinpointing this.

parafin commented:
If you manage to implement it, that would indeed be useful then.

phweyland commented:
> If you manage to implement it, that would indeed be useful then.

After checking more carefully it is like this PR work. Thanks.

phweyland commented:
I've run these test cases (seem ok to me). please add any missing case.:

new db - no auto apply - no write xmp - use embedded jpeg for thumbnail

- clean up dt files and start dt
- import 5 non xmp images, exit darkroom => history = basic(5), order none(4), V3.0(1)
- opening an image and exiting darkroom shouldn't change the figures
- write xmp for those images, remove them and import them again => basic(5), order V3.0(5)
- check altered symbol in lighttable
- remove all images

set auto-apply (all), exit & enter dt again
- import 5 non xmp images, exit darkroom => history = auto applied(1), basic(4), order none(4), V3.0(1)
- write xmp for those images, remove them and import them again => auto applied(1), basic(4), order V3.0(5)
- open one of the basic and modify the auto-applied ones. Exit darkroom => altered(2) basic(3), order V3.0(5)
- write xmp for the 2 altered images and remove them. import them again => altered(2) basic(3), order V3.0(5)
- opening a basic image and exiting darkroom should move this image to auto-applied


set don't use embedded preview jpeg and replay all
(exiting dt after every setting change & clean up xmp files)

- import 5 non xmp images, exit darkroom => history = basic(5), V3.0(5)
- write xmp for those images, remove them and import them again => basic(5), order V3.0(5)
- remove all images

set auto-apply (all), exit & enter dt again
- import 5 non xmp images, exit darkroom => history = auto applied(5), V3.0(5)
- write xmp for those images, remove them and import them again => auto applied(5), order V3.0(5)
- open two images and modify them. Exit darkroom => altered(2) auto applied(3), order V3.0(5)
- write xmp for the 2 altered images and remove them. import them again => altered(2) auto applied(3), order V3.0(5)

import of raw file with old xmp.
- xmp without history => basic or auto apply depending on setting (not tested)
- xmp with basic module only => basic
- xmp with altered images => altered
- open a folder with images with xmp (112 images) altered(108), basic(4), order custom(3), none(3), v3.0(106). 
none(3) is for image where duplicate 0 doesn't exist. basic(4) is for previuos 3 plus a png.

imported db

- clean dt files, copy dt files from backup and start dt and upgrade
- I get (from 2.6) altered(25426), basic(26973) = 52399 order custom(13), legacy(48451), none(3935) = 52399 
- check the consitency of the numbers
- open one none raw image and exit darkroom. it should move to v3.0. No change on history is no auto apply. Move to auto applied otherwise.
- open one legacy image and exit darkroom, history split should not change

TurboGit commented:
@phweyland : Not enough time to do a test and review of this at the moment. But this is not forgotten on my side.

TurboGit commented:
Really nice this new version ! I found a little glitch though.

From lightttable use the change orientation of one image and the status stays as non altered but as soon as you enter and leave darkrroom the status is updated. I suppose a missing update for the orientation change from ligthttable ?

phweyland commented:
> Really nice this new version ! I found a little glitch though.
> 
> From lightttable use the change orientation of one image and the status stays as non altered but as soon as you enter and leave darkrroom the status is updated. I suppose a missing update for the orientation change from ligthttable ?

Yes, indeed !

phweyland commented:
I've let some commented lines of code (printf) in the history_hash routines. Let me know if have to remove them.

TurboGit commented on src/libs/collect.c:
```diff
@@ -1304,21 +1304,22 @@ static void list_view(dt_lib_collect_rule_t *dr)
                    " GROUP BY altered"
                    " ORDER BY altered ASC",
-                    ("initial"), _("auto applied"), _("altered"), _("none"), where_ext);
+                   _("basic"), _("auto applied"), _("altered"), _("basic"), where_ext);
```
2 times basic here ? Is that expected?

phweyland replied:
yes, here we want to retrieve both the images without history + images which have current = basic hash into the same basic group.
Now, there is another way: to have a group "never opened" (= no history) separated from basic (current = basic hash). "Never opened" group should match "order=none". This is more factual but I'm not sure that will help the end user.
Let me know if you thinks this last way is better.

johnny-bit replied:
I'm for "never opened" being present.

parafin replied:
"never opened" stops being factual if you have disabled embedded thumnails

phweyland replied:
> "never opened" stops being factual if you have disabled embedded thumbnails

This setting is not what one changes everyday. If not set, "never opened" tells the truth (and suggests the user to open the images) otherwise it doesn't appear.
Would be "not developed"  better (more accurate but less suggestive) ?

johnny-bit replied:
"not developed" seems like better alternative, because as you've said it's more accurate.

parafin replied:
Isn't "basic" (and actually "auto applied") also "not developed"? What use-case do you have in mind? And why would you want to add a feature that only works when embedded thumbnails are enabled?

phweyland replied:
Not sure to understand your concern.

> Isn't "basic" (and actually "auto applied") also "not developed"? What use-case do you have in mind? 

As is we have:
- basic = `no history` (= no order = no hash = not developed) + `current_hash = basic_hash` (= only mandatory modules (then developed but not altered))
- auto applied = `current_hash = auto_hash ` (then develop but not altered)
- altered = current_hash != basic_hash and current_hash != auto_hash 

The discussion here is just to decide if it is worth to split the first case. We would have then:
- not developed (appears only if enabled embedded thumbnails) potentially basic or auto-applied
- basic (mandatory module only)
- auto applied (mandatory module + auto applied ones)
- altered

> And why would you want to add a feature that only works when embedded thumbnails are enabled?

The fact is the current master makes confusion about no history and not altered in particular when disabled embedded thumbnails. The goal of this PR is to solve the inconsistencies listed in #4313 , in particular those introduced by the import optimization which doesn't create the corresponding history, either masking (basic stays basic and doesn't move to altered if no change) that for the user (current state of the PR) or giving him the necessary information (never opened or not developed) to understand what does happen. 

I agree that neither "never opened" and "not developed" are perfect and can lead to inadequate interpretation. What would your proposal ?
For the time being I'd prefer slightly the first solution because the user may not be interested by the import history optimization. If it was the case he could use the "module order" collection, where he can see the "none" case (=no history...).

parafin replied:
My question is how each of the mentioned cases are useful when using darktable. Otherwise it's just exposing internal states and complicating the UI.

phweyland replied:
There is still another way: create the history at import time ! In that case the former discussion has no object any more.

johnny-bit replied:
The usage might stem from module order collection being a bit wonky due to inconsistencies around history and then history being wonky due to fact that some developments were considered developed and some not.

I have couple of usecases for history (and module order) being:
1. i want list of images i haven't developed at all (just imported)
2. i want list of images that have auto-apply presets applied since I'm most sure I messd something with them.
3. i want list of developed (or altered) images (even if that's just white balance, which previously didn't affect history/altered icon), because those are most likely candidates for export
4. i want list of images with custom module order, because those are most likelly caused by my (wonky) mouse dragging stuff and in 99% of cases those might be messed edits and i need to fix those
5. i want list of images with older moduler order, so i could probably upgrade those to new order.

all those use cases and more are covered by changes by @phweyland :)

parafin replied:
So we have:

1. just imported
2. with just auto-applied presets
3. altered

and

a. old module order
b. new default module order
c. modified module order

I don't see here "none"/"no history"/"not developed" or whatever.

phweyland replied:
>     1. just imported

This is not correct.

parafin replied:
What is not correct? From user point of view there are 3 actions one can do:
1. import
2. open in darkroom
3. edit (change settings)

What else is there?

phweyland replied:
@TurboGit 
I don't see any @parafin points which would be misaligned with the current state of the PR.

I feel he is concerned with the split of basic group into "never opened" or "not developed" and "basic" one, but this is unclear for me.

Anyway, both ways are fine for me. Please let me know how to follow up on this.

parafin replied:
I'm just advocating for having some high-level design notes (e.g. usage scenarios we want to cover) and working at implementing them. The fact that "embedded thumbnails" preference affects history creation isn't something we want to expose to users. Simple way of thinking about this is to imagine how you would describe a feature in user documentation. How do you explain the difference between "never opened" and "basic" without it becoming overly complicated?

TurboGit replied:
@parafin: You missed one action from user's point of view, the fact that some auto-applied presets have been triggered or not. This is really a user's oriented action as auto-presets are added by users. So to me I still see 3 possibles states from user's point of view, but those:

1. imported (with possibly some required iop set like temperature for RAW) 
2. with auto-applied presets
3. manually edited

As discussed before (IIRC) the fact that we go into darkrroom is not an action and should not be counted if there is no modification done. Now the part that may be confusing is that we are also trying to mix information about iop-order (legacy or v3.0) and this order can be set by points 2 and point 3.

I think imported state above is what you said to be "basic", right?

parafin replied:
I'm OK with this classification, but it's still only 3 states, not 4;)

TurboGit replied:
@parafin ; If you're referring to "never opened" and "never developed" I agree with you. This won't make sense to end user's I suppose.

@phweyland : Most of the usages I see (and let's concentrate only on this for this PR, we'll see later if we want to split status) is to select quickly the non-edited images, the ones that need some work to be done. And of course the status altered from lighttable which is the "manually edited" (in darkroom or using copy/paste or style) state described above.

I think we are at this stage currently, right?

phweyland replied:
> I think we are at this stage currently, right?

Yes, that's the current stage (3 states).

phweyland replied:
> I'm just advocating for having some high-level design notes (e.g. usage scenarios we want to cover) and working at implementing them. The fact that "embedded thumbnails" preference affects history creation isn't something we want to expose to users. Simple way of thinking about this is to imagine how you would describe a feature in user documentation. How do you explain the difference between "never opened" and "basic" without it becoming overly complicated?

I agree with that. Saying that at the beginning would simplify the conversation, wouldn't it ?

TurboGit replied:
So let me test this again as it is now and merge if all goes well.

TurboGit commented:
And yes I would remove the debug printf. Thanks.

phweyland commented:
@TurboGit 
Maybe another issue for this pr. 
Some users have error on the INSERT ... ON CONFLICT query.
https://discuss.pixls.us/t/sqlite3-error-when-moving-from-lighttable-to-darkroom/16781

I think this is because there sqlite version is too old (<3.24).

I haven't not found any constraint in the current dt build.

There are 2 possibilities. Either come back to a code compatible with older sqlite version (multiple queries, slower). Or put the version constraint into the dt build.

What do you think ?

TurboGit commented:
Well, SQLite3 v3.24 has been released on 2018-06-04. So a release before that is almost two years old. That's already an issue in my point of view.

Now I'm wondering if we could replace the "ON CONFLICT" statements with "INSERT OR REPLACE" ? I'm not sure about the semantic when not all fields are specified...

phweyland commented:
> Now I'm wondering if we could replace the "ON CONFLICT" statements with "INSERT OR REPLACE" ? I'm not sure about the semantic when not all fields are specified...

The doc says they are equivalent. So I guess we will have the same release issue.

TurboGit commented:
I don't think we will have the same issue as we have "INSERT OR REPLACE"  statements in current code since long time. So the  "ON CONFLICT" is probably just a new supported syntax. to be tested

phweyland commented:
I've made the confusion between INSERT OR REPLACE and INSERT OR UPDATE.
Anyway I'll make a try with INSERT OR REPLACE to see if if that accepts a subset of columns.

phweyland commented:
>  I'll make a try with INSERT OR REPLACE to see if if that accepts a subset of columns.

Works nicely. Why to do simple when we can do complicate ???
I'll issue a PR.

phweyland commented:
No doesn't work.
a not null field is lost when it is not provided by INSERT OR REPLACE.

phweyland commented:
UPSERT (UPDATE OR INSERT) has been introduced at the same time as ON CONFLICT.

johnny-bit commented:
I'd favour bumping (and naming verbatimly) sqlite requirement to 3.24.

It seems that all major distros have it in proper ver based on repology: https://repology.org/project/sqlite3/versions

TurboGit commented:
@johnny-bit : I agree and I'll do that.

phweyland commented:
> It seems that all major distros have it in proper ver based on repology: https://repology.org/project/sqlite3/versions

I don't understand what tells this site ...

Some user says that it's not available on its distribution.
A third way would be to provide conditional code before/after 3.24 to keep the compatibility.
Do you want me to do that ?

johnny-bit commented:
> I don't understand what tells this site ...

It's just list of package versions in various distro/version. From what I see, worst offenders and eadaches will be Ubuntu 18.04, since it uses 3.22

> A third way would be to provide conditional code before/after 3.24 to keep the compatibility.
> Do you want me to do that ?

While it'd be cool, I think that your history-related work will land in "stable" release 3.2 in december, and by that time sqlite 3.24 will be 2.5 years old, so requiring "sqlite ver >= 3.24" won't be seen as an outrageous requirement ;)
